kernel/async/monitor-common/async/TaskletQueue.hh

```diff
@@ -28,108 +28,155 @@
-  TaskletBase* oldtail = sharedTail.exchange(t, std::memory_order_release); // replace the tail
-  t->nextTasklet.store(oldtail, std::memory_order_relaxed); // comlete the push
-  return oldtail == FREE(); // true if this was the first message in the queue
-}
-
-inline Tasklet* TaskletQueue::pull() {
-  // try to retrieve from the private queueu 
-  TaskletBase* oldtail = privateTail.exchange(FREE(), std::memory_order_relaxed); // avoid load()
-  ASSERT(oldtail != LOCKED());
-  if (oldtail != FREE()) {
-    privateTail.store(oldtail->nextTasklet.load(std::memory_order_relaxed),
-		      std::memory_order_relaxed); // remove oldtail from private queue
-    return static_cast<Tasklet*>(oldtail);
-  } // else try to retrieve from the shared queue
-  oldtail = sharedTail.exchange(LOCKED(), std::memory_order_relaxed); // detach tail and mark as locked
-  if (oldtail == FREE() || oldtail == LOCKED()) return nullptr; // nothing was in the shared queue, but now it is locked
-  while (!(oldtail == FREE() || oldtail == LOCKED())) {
-    TaskletBase* next;
-    do { // wait until the push is no longer marked as incomplete
-      // use exchange in order to avoid shared state of cacheline
-      next = oldtail->nextTasklet.exchange(INCOMPLETE(), std::memory_order_acquire);
-      if (next == INCOMPLETE()) hwthread_pause(); // sleep for a short amount of cycles
-    } while (next == INCOMPLETE());
-    if (next == FREE() || next == LOCKED()) {
-      return static_cast<Tasklet*>(oldtail); // directly return last task from old shared queue
-    } // else push it into the private queue
-    oldtail->nextTasklet.store(privateTail.exchange(oldtail, std::memory_order_relaxed),
-			       std::memory_order_relaxed);
-    oldtail = next;
-  }
-  PANIC_MSG(false, "should never reach here");
-}
-
-inline bool TaskletQueue::tryRelease()
-{
-  ASSERT(privateTail == FREE());
-  TaskletBase* oldtail = LOCKED();
-  return sharedTail.compare_exchange_strong(oldtail, FREE(), std::memory_order_relaxed);
-}
-
-} // async
-} // mythos
+
+  class TaskletQueueBaseDefault {
+  public:
+    TaskletQueueBaseDefault()
+      :privateTail_(Tasklet::FREE), sharedTail_(Tasklet::FREE)
+    {}
+
+    std::atomic<uintptr_t>& privateTail() {
+      return privateTail_;
+    }
+
+    std::atomic<uintptr_t>& sharedTail() {
+      return sharedTail_;
+    }
+
+  private:
+    std::atomic<uintptr_t> privateTail_;
+    std::atomic<uintptr_t> sharedTail_;
+  };
+
+  class TaskletQueueBaseAligned {
+  public:
+    TaskletQueueBaseAligned()
+      :privateTail_(Tasklet::FREE), sharedTail_(Tasklet::FREE)
+    {}
+
+    std::atomic<uintptr_t>& privateTail() {
+      return privateTail_;
+    }
+
+    std::atomic<uintptr_t>& sharedTail() {
+      return sharedTail_;
+    }
+
+  private:
+    alignas(64) std::atomic<uintptr_t> privateTail_;
+    alignas(64) std::atomic<uintptr_t> sharedTail_;
+  };
+
+  template<typename BASE>
+  class TaskletQueueImpl
+  {
+  private:
+    BASE base;
```

Comment:
I would do this via a protected base class. Its more compact, especially when called "base" anyway. Then `base.sharedTail()` becomes `this->sharedTail()`.